### PR TITLE
TypeScript example proposal

### LINKED ISSUE #96: Add typescript example
this looks _awesome_ but I also really like TypeScript. 
is there any way to integrate typescript into the workflow?

### PR DESCRIPTION
**Pull request by giacomorebonato**
Added a very basic example using TypeScript.
Livereload seems to work.

To test it just run  
```
npm install  
npm run dev  
```

Using `concurrently` to run both next and TypeScript compiler with one command.

Fixes #96

### CODE REVIEW

File: examples/with-typescript/typings.json
Comment (giladgray) [inline examples/with-typescript/typings.json]: i would suggest using `@types/react` directly in NPM instead of relying on `typings`. 
it's the way of the future, and you wouldn't need `postinstall` task.

File: examples/with-typescript/package.json
Comment (giladgray) [inline examples/with-typescript/package.json]: shouldn't need to add `$(npm bin)/` in package.json scripts--NPM configures `PATH` automatically for you when running npm scripts.

should be safe to just say `"concurrently 'tsc --watch' 'next'"`

File: examples/with-typescript/package.json
Comment (giladgray) [inline examples/with-typescript/package.json]: i'd argue that typescript is a `devDependency`

**Review by giladgray — CHANGES_REQUESTED** [review]
this is a great minimal example! just a few suggestions for a simpler setup.

File: examples/with-typescript/typings.json
Comment (giacomorebonato) [inline examples/with-typescript/typings.json]: Thanks,  didn't know!

Comment (giacomorebonato) [inline examples/with-typescript/typings.json]: mmmh I have tried, but I now am receiving many errors when compiling TypeScript.

Comment (giacomorebonato) [inline examples/with-typescript/typings.json]: My fault... I didn't remove previous typings.

**Comment by jagreehal** [comment]
Hey @giacomorebonato I'm trying a similar experiment see https://github.com/jagreehal/next-typescript . You're going to have the problem of the page refreshing twice until Next v2.1

File: examples/with-typescript/typings.json
Comment (rauchg) [inline examples/with-typescript/typings.json]: If you make this change I'll merge!

**Comment by OskarKlintrot** [comment]
This wont be able to handle static content either if you want to separate the build and source. I will try to create a simple example using tscomp for the build process, which hopefully will be really neat!

File: examples/with-typescript/typings.json
Comment (giacomorebonato) [inline examples/with-typescript/typings.json]: it's there! Is it what you mean right?

**Comment by giacomorebonato** [comment]
Looking forward to it @OskarKlintrot

**Review by timneutkens — COMMENTED** [review]
Cool, though I wonder if we can use webpack here 🤔

**Comment by giacomorebonato** [comment]
A setup like this works pretty well with tests without webpack. 
However it is a good topic.

**Comment by giacomorebonato** [comment]
As a Typescript enthusiast I feel freedom in just using the native compiler, without clashing on other JS build processes.

File: examples/with-typescript/package.json
Comment (rauchg) [inline examples/with-typescript/package.json]: A concern is that this will get out of sync with next's react

Comment (giacomorebonato) [inline examples/with-typescript/package.json]: I have remove React dependency. I thought it was a peer one.

File: examples/with-typescript/package.json
Comment (rauchg) [inline examples/with-typescript/package.json]: This doesn't seem to depend on a specific version of next. Can you use `latest` here?

Comment (giacomorebonato) [inline examples/with-typescript/package.json]: Sure.

File: examples/with-typescript/package.json
Comment (rauchg) [inline examples/with-typescript/package.json]: Why the babel preset? Is that a TypeScript requirement?

**Comment by OskarKlintrot** [comment]
Now I have an example of using `tscomp` with next [here](https://github.com/oskarklintrot/next-typescript). It's based on @jagreehal example. However, css modules and images are not working, so this is more or less the same example as @jagreehal's but with an overkill build process in place (tscomp is built on top of create-react-app but also work for server apps). I do would love to just use the native compiler @giacomorebonato but as long as it's not working for static assets without building the js to the same dir as the ts then it is a bit limited. To me personally it would be a deal breaker to have the src and build in the same place. I will keep looking into setting up tscomp to work with the static assets as well.

File: examples/with-typescript/package.json
Comment (giacomorebonato) [inline examples/with-typescript/package.json]: I thought I was getting an error without it... but I have tried again and it's not needed.
Removed.

**Comment by jagreehal** [comment]
@OskarKlintrot nice I'll take a look. I've also been thinking about adding a babel-loader!ts-loader here  https://github.com/zeit/next.js/blob/master/server/build/webpack.js

**Comment by giacomorebonato** [comment]
In VSCode it's possible to hide the .js files and see only the .ts ones.  
I like this solution, because it allows to use next.js simple as it is.

**Review by giladgray — APPROVED** [review]
this seems like an ideal minimal typescript example: it provides a clean solution for the __single problem of typescript compilation__. it does not presume to handle more than it needs to, such as testing (solved by dozens of other tools; i suggest [ts-jest](https://github.com/kulshekhar/ts-jest)) or static assets (supported by [next itself](https://github.com/zeit/next.js#static-file-serving-eg-images)).

File: examples/with-typescript/tsconfig.json
Comment (OskarKlintrot) [inline examples/with-typescript/tsconfig.json]: Why not using source maps?

Comment (giacomorebonato) [inline examples/with-typescript/tsconfig.json]: You are right, they are useful. Will change this

Comment (giladgray) [inline examples/with-typescript/tsconfig.json]: might be best to make this file as simple as possible, the bare minimum to compile successfully. let other folks choose their compiler flags as necessary. 

just set `jsx`, `module`, and `target`.

Comment (OskarKlintrot) [inline examples/with-typescript/tsconfig.json]: Makes sense!

Comment (OskarKlintrot) [inline examples/with-typescript/tsconfig.json]: (and maybe change `es6` to `es2015` while your at it ;-) )

**Comment by OskarKlintrot** [comment]
Just did a PR to @giacomorebonato fork :)

**Comment by giacomorebonato** [comment]
hi @OskarKlintrot I have seen you PR.  
I don't see `.gitignore` in the other examples.  
I am willing to convert one component to a functional one, but I will keep the other extending `React.Component`, since this is not a boilerplate but just a small example. ;)

**Comment by OskarKlintrot** [comment]
It's bad practice to version control build files, that's the reason behind
gitignore. Ignoring *.js will probably not be a good idea in the other
examples...

Using stateless functions is [the preferred way of doing it in react](
https://facebook.github.io/react/blog/2015/09/10/react-v0.14-rc1.html#stateless-function-components)
so why do something different in the example?

Den mån 23 jan. 2017 18:51Giacomo Rebonato <notifications@github.com> skrev:

**Comment by giacomorebonato** [comment]
ah ignoring `.js` it is a good idea sorry if I read the PR too fast.  
I will accept it and update this.  
The reason why I used a stateful component is because it is used also in next.js readme for showing `getInitialProps` usage, but I can change it too.

**Comment by giacomorebonato** [comment]
@OskarKlintrot I just added another commit. In the gitignore I have considered only the .js files since .next and node_modules are excluded by top level .gitignore.

**Comment by OskarKlintrot** [comment]
You really kill all joy and motivation of helping out code wise when you take my changes but ignore my commits (PR), thous taking away all the credits for what I have actually done and leaving me out as a contributor.

**Comment by rauchg** [comment]
Maybe it was a tooling issue! Would love to have you both as contributors to have the best possible TypeScript example 😇

**Comment by OskarKlintrot** [comment]
Thank you @rauchg, really appreciating it!

I still see stuff to improve on (this example won't run on Windows, at least not on PS, for example), maybe I can calm myself down a bit and do a new PR later today.

Changing subject, this commit history is really growing out of proportion, 13 commits and counting! The commits are longer than the code! 🙀

**Comment by giacomorebonato** [comment]
@OskarKlintrot sorry... I didn't want to disrespect you.  I was just hoping to have this merged soon and it was a rookie mistake. Sorry.  
If you resolve the conflicts in your PR I will merge it now if I am still in time.

**Comment by giacomorebonato** [comment]
I was trying to squash the commits... and I did a mess, trying to repair.

**Comment by OskarKlintrot** [comment]
The commit history looks much better! I can make a new PR in a couple of minutes!

**Comment by giacomorebonato** [comment]
merged!

File: examples/with-typescript/package.json
Comment (timneutkens) [inline examples/with-typescript/package.json]: Does `next` need to be quoted? Since it doesn't have any options. 🤔

Comment (OskarKlintrot) [inline examples/with-typescript/package.json]: Good question! I guess not, hang on!

Comment (OskarKlintrot) [inline examples/with-typescript/package.json]: Nope, not needed.

Comment (albinekb) [inline examples/with-typescript/package.json]: @OskarKlintrot You could rebase this into the other "cleanup" commit :1st_place_medal:

Comment (OskarKlintrot) [inline examples/with-typescript/package.json]: True! It's my commit but on @giacomorebonato fork that I don't have push access to, can @giacomorebonato rebase it or do I have to do it and make a new PR to that fork? This is a bit outside my comfort zone with git.

Comment (giacomorebonato) [inline examples/with-typescript/package.json]: I will try to make the change and then rebase.

**Comment by giacomorebonato** [comment]
I have noticed problems when using `next/head` with this configuration.  
The head tags are taken, but a warning is thrown up with a message about trying to render an empty component.

**Comment by timneutkens** [comment]
What a huge PR has this gotten into 😅, I'll check that error @giacomorebonato

**Comment by timneutkens** [comment]
Also, there is no README.md 😥

**Comment by giacomorebonato** [comment]
Hi @timneutkens I just came back from holidays. I will update this PR soon with a README.md too!

File: examples/with-typescript/README.md
Comment (timneutkens) [inline examples/with-typescript/README.md]: Typo:

>In `tsconfig.json` the option `"jsx": "react"` compiles JSX syntax into nested React.createElement calls.

**Review by timneutkens — APPROVED** [review]
Looks good 👍

**Comment by timneutkens** [comment]
Fixes #96

**Comment by giacomorebonato** [comment]
Thanks! I hope to investigate more and to help!

**Comment by karthikiyengar** [comment]
Hi @giacomorebonato - How do you get sourcemaps to work reliably here? Runtime errors point to the transpiled JS code instead of the ts source files. Is there any workaround? I've tried enabling sourceMaps in tsconfig.

**Comment by giladgray** [comment]
@karthikiyengar i'm no expert in next's webpack setup, but typically you need to add `source-map-loader` to load source maps from externally compiled files. have you tried mucking with `next.config.js`?

**Comment by giacomorebonato** [comment]
Latest release of TypeScript compiles JSX files to JS. I want to try to edit the boilerplate in a way that we rely only on webpack for compilation and this should help also for the sourcemaps.

**Comment by bestwestern** [comment]
I have mad a pull request where tsx compiles directly to: JS https://github.com/zeit/next.js/pull/1600

**Comment by jagreehal** [comment]
I wrote a game for a talk on Typed JavaScript and used TypeScript and Next
- https://github.com/jagreehal/typescript-react-redux-jest-card-game

Hopefully this helps.

**Pull request merged**
